src/permutations.rs

```diff
@@ -41,34 +51,44 @@ enum PermutationState {
     End,
 }
 
-impl<I> fmt::Debug for Permutations<I>
+impl<I, Item> fmt::Debug for PermutationsGeneric<I, Item>
 where
     I: Iterator + fmt::Debug,
     I::Item: fmt::Debug,
 {
     debug_fmt_fields!(Permutations, vals, state);
 }
 
+pub fn array_permutations<I: Iterator, const K: usize>(iter: I) -> ArrayPermutations<I, K> {
+    PermutationsGeneric {
+        vals: LazyBuffer::new(iter),
+        state: PermutationState::Start { k: K },
+        _item: PhantomData,
+    }
+}
+
 pub fn permutations<I: Iterator>(iter: I, k: usize) -> Permutations<I> {
-    Permutations {
+    PermutationsGeneric {
         vals: LazyBuffer::new(iter),
         state: PermutationState::Start { k },
+        _item: PhantomData,
     }
 }
 
-impl<I> Iterator for Permutations<I>
+impl<I, Item> Iterator for PermutationsGeneric<I, Item>
 where
     I: Iterator,
     I::Item: Clone,
+    Item: PermItem<I::Item>,
 {
-    type Item = Vec<I::Item>;
+    type Item = Item;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let Self { vals, state } = self;
+        let Self { vals, state, _item } = self;
         match state {
             PermutationState::Start { k: 0 } => {
                 *state = PermutationState::End;
-                Some(Vec::new())
+                Some(Item::extract_start(vals, 0, 0))
```

Comment:
I see calling `extract_start` with 0 here has the merit to save one function, but it requires us to account for the special case `len==0` in `extract_start`, right? Would unconditionally working on `indices` simplify and get rid of special cases?

Reply:
Correct, this attempt won't work. However, we could introduce `PoolIndex::from_fn` to do whatever we want and return either a `Vec` or an array.